### DISCUSSION TITLE
adjust reference text @eq -> Eq. (X) ; @fig -> Fig. X ; @section -> Sec. X

### DISCUSSION THREAD

**Discussion by Qottmann** [typst/typst#2243] [Q&A] [answered]
Hello,

I want to adjust the displayed reference text (is there a name for that?) such that @eq -> Eq. (X) ; @fig -> Fig. X ; @section -> Sec. X. I have a working example

https://typst.app/project/rVhSwwqM3Z412KA3UJrd14

But this is quite cumbersome. Is there an easier way to do this without doing 3 show rules?

```typst
#show ref: it => {
  if it.element != none and it.element.func() == figure {
    link(it.target)[Fig.~#it.element.numbering]
  } else {
    it
  }
}
```

edit: Just noticed I have a bug and all figures are referenced as Fig. 1, so doesnt work quite yet. I guess `#it.element.numbering` is taking the numbering style, and not the numbering of that instance 🤔 How do I get that instead?

edit: `#it.element.counter.display()` starts counting from 0 🙄

**Comment by xkevio** [comment, marked as answer]
I think besides the equation part, everything else can just be set-rules, no?
This does mean that the reference supplement for `figure` is the same as the caption text.

For your current solution, you can put this all into the same show-rule and differentiate with the `if`,
might be a bit nicer than 3 separate ones.

```typ
#show ref: it => {
  if it.element != none and it.element.func() == math.equation {
    link(it.target)[#it.element.supplement #numbering(it.element.numbering, ..counter(math.equation).at(it.element.location()))]
  } else {
    it
  }
}

#set heading(numbering: "1", supplement: "Sec.")
#set math.equation(numbering: "(1)", supplement: "Eq.")
#set figure(supplement: "Fig.")

= hallo<a>
@a 

$ x + y $<y>

See @y.

#figure(caption: "test", [])<fig>
See @fig.
```

**Reply by Qottmann** [reply]
Thanks so much, exactly what I was looking for!

**Reply by laurmaedje** [reply]
Note that `counter(math.equation).display()` will display the counter value at the reference, not at the equation. `numbering(it.element.numbering, ..counter(math.equation).at(it.element.location())` or something like that would give the correct number.

**Reply by Qottmann** [reply]
"will display the counter value at the reference, not at the equation" not sure I understand, can you elaborate?
Edit: Ah okay this is always displaying the number where the counter currently is at, in the text.

**Reply by Qottmann** [reply]
@xkevio can you updated the answer marked as solution to
```typst
#show ref: it => {
  if it.element != none and it.element.func() == math.equation {
    link(it.target)[#it.element.supplement #numbering(it.element.numbering, ..counter(math.equation).at(it.element.location()))]
  } else {
    it
  }
}

#set heading(numbering: "1", supplement: "Sec.")
#set math.equation(numbering: "(1)", supplement: "Eq.")
#set figure(supplement: "Fig.")

= hallo<a>
@a 

$ x + y $<y>

See @y.

#figure(caption: "test", [])<fig>
See @fig.
```
